serverv2/appmanager/stf_router.go

```diff
@@ -0,0 +1,98 @@
+package appmanager
+
+import (
+	"context"
+	"errors"
+	"fmt"
+)
+
+var (
+	ErrNoHandler = errors.New("no handler")
+)
+
+// MsgHandler is a function that handles the message execution. TODO: move to appmanager.Module.go
+type MsgHandler = func(ctx context.Context, msg Type) (msgResp Type, err error)
+
+// PreMsgHandler is a function that executes before the message execution.TODO: move to appmanager.Module.go
+type PreMsgHandler = func(ctx context.Context, msg Type) (err error)
+
+// PostMsgHandler is a function that executes after the message execution.TODO: move to appmanager.Module.go
+type PostMsgHandler = func(ctx context.Context, msg Type, msgResp Type) (err error)
+
+type QueryHandler = MsgHandler
+
+func newMsgRouterBuilder() *msgRouterBuilder {
+	return &msgRouterBuilder{
+		handlers:     make(map[string]MsgHandler),
+		preHandlers:  make(map[string][]PreMsgHandler),
+		postHandlers: make(map[string][]PostMsgHandler),
+	}
+}
+
+type msgRouterBuilder struct {
+	handlers     map[string]MsgHandler
+	preHandlers  map[string][]PreMsgHandler
+	postHandlers map[string][]PostMsgHandler
+}
+
+func (b *msgRouterBuilder) RegisterHandler(msgType string, handler MsgHandler) error {
+	// panic on override
+	if _, ok := b.handlers[msgType]; ok {
+		return fmt.Errorf("handler already registered: %s", msgType)
+	}
+	b.handlers[msgType] = handler
+	return nil
+}
+
+func (b *msgRouterBuilder) RegisterPreHandler(msgType string, handler PreMsgHandler) {
+	b.preHandlers[msgType] = append(b.preHandlers[msgType], handler)
+}
+
+func (b *msgRouterBuilder) RegisterPostHandler(msgType string, handler PostMsgHandler) {
+	b.postHandlers[msgType] = append(b.postHandlers[msgType], handler)
+}
+
+func (b *msgRouterBuilder) Build() (MsgHandler, error) {
+	handlers := make(map[string]MsgHandler)
+	for msgType, handler := range b.handlers {
+		// find pre handler
+		preHandlers := b.preHandlers[msgType]
+		// find post handler
+		postHandlers := b.postHandlers[msgType]
+		// build the handler
+		handlers[msgType] = buildHandler(handler, preHandlers, postHandlers)
+	}
+	// TODO: add checks for when a pre handler/post handler is registered but there is no matching handler.
+
+	// return handler as function
+	return func(ctx context.Context, msg Type) (Type, error) {
+		typeName := TypeName(msg)
+		handler, exists := handlers[typeName]
+		if !exists {
+			return nil, fmt.Errorf("%w: %s", ErrNoHandler, typeName)
+		}
+		return handler(ctx, msg)
+	}, nil
+}
+
+func buildHandler(handler MsgHandler, preHandlers []PreMsgHandler, postHandlers []PostMsgHandler) MsgHandler {
+	// TODO: maybe we can optimize this by doing a switch case and checking if the pre/post handlers are empty
+	// in order to avoid pointless iterations when there are no pre/post handlers
```

Comment:
since this is only at startup, its less of a worry. If the optimisation keeps the code legible we can move to it, otherwise this is easy to understand